spec.md

```diff
@@ -3149,6 +3167,227 @@ The CO MUST implement the specified error recovery behavior when it encounters t
 | Snapshot list mismatch | 3 INVALID_ARGUMENT | Besides the general cases, this code SHOULD also be used to indicate when plugin supporting CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT detects a mismatch in the `snapshot_ids`. | If a mismatch is detected in the `snapshot_ids`, caller SHOULD use different `snapshot_ids`. |
 | Volume group snapshot does not exist | 5 NOT_FOUND | Indicates that a volume group snapshot corresponding to the specified `group_snapshot_id` does not exist. | Caller MUST verify that the `group_snapshot_id` is correct and that the volume group snapshot is accessible and has not been deleted before retrying with exponential back off. |
 
+### Snapshot Metadata Service RPCs
+
+The Snapshot Metadata service is an optional service that is used to retrieve metadata on the allocated blocks of a single snapshot, or the changed blocks between a pair of snapshots of the same block volume.
+Retrieval of the data blocks of a snapshot is not addressed by this service and it is assumed that existing mechanisms to read snapshot data will suffice.
+
+#### Metadata Format
+
+Block volume data ranges are specified by a sequence of `BlockMetadata` tuples.
+Tuples in a sequence are in ascending order of `byte_offset`, with no overlap between adjacent tuples.
+
+```protobuf
+
+// BlockMetadata specifies a data range.
+message BlockMetadata {
+  // This is the 0 based byte position in the volume or snapshot,
+  // measured from the start of the object.
+  // This field is REQUIRED.
+  uint64 byte_offset = 1;
+
+  // This is the size of the data range.
+  // size_bytes MUST be greater than 0.
+  // This field is REQUIRED.
+  uint64 size_bytes = 2;
+}
+
+```
+
+The `BlockMetadataType` enumerated type describes how the `size_bytes` fields of `BlockMetadata` tuples may vary in a sequence. There are two prevalent styles in which data ranges are described:
+
+- The **FIXED_LENGTH** style requires a fixed value for the `size_bytes` field in all tuples in a given sequence.
+- The **VARIABLE_LENGTH** style does not constrain the value of the `size_bytes` field in a sequence.
+
+The Snapshot Metadata service permits either style at the discretion of the plugin as long as the style does not change mid-stream in any given RPC.
```

Comment:
this seems to imply that subsequent invocations of the same RPC may specify different length styles - even for the same snapshot. is this intended? i hope so, because otherwise the plugin would need to maintain state to enforce consistent styles across multiple RPCs - and we probably don't want to force plugins to do that

Reply:
Yes. The spec leaves it to the SP to chose the style on a per RPC basis; consistency is required only within each RPC.